docs/src/main/asciidoc/security-oidc-code-flow-authentication-concept.adoc

```diff
@@ -107,104 +378,224 @@ public class ProtectedResource {
+If you expect the access token to contain the roles required to access the current Quarkus endpoint (`quarkus.oidc.roles.source=accesstoken`) then it will also be verified.
+--
 
 [[token-introspection-userinfo-cache]]
-=== Token introspection and UserInfo cache
+==== Token introspection and UserInfo cache
 
-Code flow access tokens are not introspected unless they are expected to be the source of roles but will be used to get `UserInfo`. So there will be one or two remote calls with the code flow access token, if the token introspection and/or `UserInfo` are required.
+Code flow access tokens are not introspected unless they are expected to be the source of roles.
+It will however be used to get `UserInfo`.
+There will be one or two remote calls with the code flow access token, if the token introspection and/or `UserInfo` are required.
 
 Please see xref:security-oidc-bearer-token-authentication-concept.adoc#token-introspection-userinfo-cache[Token Introspection and UserInfo cache] for more information about using a default token cache or registering a custom cache implementation.
 
 [[jwt-claim-verification]]
-=== JSON web token claim verification
+==== JSON web token claim verification
 
 Please see xref:security-oidc-bearer-token-authentication-concept.adoc#jwt-claim-verification[JSON Web Token Claim verification] section about the claim verification, including the `iss` (issuer) claim.
 It applies to ID tokens but also to access tokens in a JWT format if the `web-app` application has requested the access token verification.
 
-=== Redirection
+==== Further security with Proof key for code exchange (PKCE)
 
-When the user is redirected to the OpenID Connect Provider to authenticate, the redirect URL includes a `redirect_uri` query parameter which indicates to the provider where the user has to be redirected to once the authentication has been completed.
+link:https://datatracker.ietf.org/doc/html/rfc7636[Proof Key for Code Exchange] (PKCE) minimizes the risk of the authorization code interception.
 
-Quarkus will set this parameter to the current request URL by default. For example, if the user is trying to access a Quarkus service endpoint at `http://localhost:8080/service/1` then the `redirect_uri` parameter will be set to `http://localhost:8080/service/1`. Similarly, if the request URL is `http://localhost:8080/service/2` then the `redirect_uri` parameter will be set to `http://localhost:8080/service/2`, etc.
+While `PKCE` is of primary importance to the public OpenID Connect clients (such as the SPA scripts running in a browser), it can also provide an extra level of protection to Quarkus OIDC `web-app` applications which are confidential OpenID Connect clients capable of securely storing the client secret and using it to exchange the code for the tokens.
```

Comment:
```suggestion
While `PKCE` is of primary importance to public OpenID Connect clients, such as SPA scripts running in a browser, it can also provide an extra level of protection to Quarkus OIDC `web-app` applications. 
Quarkus OIDC `web-app` applications are confidential OpenID Connect clients capable of securely storing the client secret and using it to exchange the code for the tokens.
```

Reply:
@michelle-purcell It feels like that describing that `web-app` applications are confidential clients in a separate sentence kind of loses the link with the `PKCE` in this context. Looking at again, I wonder if we can just replace `which` with `even though` to make it clearer that PKCE is not strictly essential for this type of the application:
```
While `PKCE` is of primary importance to public OpenID Connect clients, such as SPA scripts running in a browser, it can also provide an extra level of protection to Quarkus OIDC `web-app` applications even though they act as confidential OpenID Connect clients capable of securely storing the client secret and using it to exchange the code for the tokens.
```

What do you think ?

Reply:
I updated my suggestion above. WDYT?